mlx/fast.cpp

```diff
@@ -94,11 +97,71 @@ array rms_norm(
   return fallback({x, weight})[0];
 }
 
+std::vector<array> RMSNorm::vjp(
+    const std::vector<array>& primals,
+    const std::vector<array>& cotangents,
+    const std::vector<int>& argnums,
+    const std::vector<array>& outputs) {
+  assert(primals.size() == 2);
+  assert(outputs.size() == 1);
+  assert(cotangents.size() == 1);
+
+  auto s = stream();
+  auto fallback = [eps = eps_, s](const std::vector<array>& inputs) {
```

Comment:
Unrelated question, do you think we should implement these in terms of the VJP of the original fallback?

Reply:
Yeah I did it for efficiency purposes indeed. I first tested with just the fallback and then implemented the kernel. It was the natural order of things anyway since I first did the vjp by hand then wrote it in normal code, tested that it is correct (iterate these two steps 50 times, damn it had been a while since I computed complicated gradients) then wrote it in a kernel.

I was thinking what would be a good way to automatically test those. Perhaps vmap a singleton dimension in the beginning?

Reply:
>I was thinking what would be a good way to automatically test those. Perhaps vmap a singleton dimension in the beginning?

CPU?

Reply:
Maybe we do test them already when we run the CPU tests actually 🤔

Reply:
Yea a vmap over the VJP might work. Or just a vjp of the vjp.

Reply:
Yeah exactly. Vmap causes problems cause scatter lacks a vmap (which we should fix).

Reply:
Hmm that reminds me of a discussion I forgot about 😬